### ISSUE TITLE
Fixed: Incorrect context data passed by `LayoutRelationshipAuthorFromNode`.

### ISSUE THREAD

**Issue by docwilmot** [backdrop/backdrop-issues#5482]
## Description of the bug

The `LayoutRelationshipAuthorFromNode` plugin is passing a User context with Node data.

## Steps To Reproduce

Create a Node layout, (path `node/%`). Add a `Author from node` relationship. The context in the layout will be seen to be a User context as expected, but the `data` value in this so-called user context will be the Node in position 2. This can be demonstrated by either doing a `dpm($this->contexts)` in `layout.class.inc`, or more practically, create a block with a required context of `user => user`:
```
/**
 * Implements hook_block_info().
 */
function tester_block_info() {
  $blocks = array();
  $blocks['test_block'] = array(
    'info' => t('test'),
    'description' => t('test'),
    'required contexts' => array(
      'user' => 'user',
    ),
  );

  return $blocks;
}

/**
 * Implements hook_block_view().
 */
function tester_block_view($delta = '', $settings = array(), $contexts = array()) {
  $block = array();
  dpm($contexts);
  switch ($delta) {
    case 'test_block':
      $block['content'] = 'test';
      $block['subject'] = t('test');
      break;
  }
  return $block;
}
```
Then navigate to any node.

## Actual behavior

The `dpm()` on block view (the whole point of context and relationships really) will return a User context, but with a Node object as its data, the node from position 2. 

## Expected behavior

Blocks should get the Author of the node as the object in context data. 

## Additional information

1.22.0

**Comment by docwilmot** [comment]
This one is a bit unsatisfactory I admit. The problem here was/is that when a page loads, `layout_route_handler()` calls `layout_get_layout_by_path()` to get the layout, which immediately seeks to set contexts by calling `getContexts()` on the layout, and only then, after the contexts array is built, can the menu object data (node here) be loaded into any path-related contexts (like if youre on a `node/%` page for example). But this means that the first time `getContexts()` is called on a page with a Node based on the path, there is no node data (yet) to pass on to the relationship plugins that create a context from object data, like User from Node.

Incidentally, the code was creating a User context with empty data instead and `layout_get_layout_by_path()` was simply sticking a node context in there, hence this issue. This PR stops this; if the User from Node plugin doesnt get a context with existing Node data, it will now just not create a new User context. 

But this means no User from Node if Node comes from a path context.

To fix the issue then, a few options:
- add another call to `getContexts()` after `layout_get_layout_by_path()` has just finished building the context array and loading it with a path object. This is what this PR does. Aside, since another PR awaiting review seeks to cache `getContexts()`, this should eventually not be a problem.
- make `Layout::getContextsFromRelationships()` a public function and call that instead of a repeat `getContexts()`
- pass the path to `getContexts()` so it can load the data itself instead of `layout_get_layout_by_path()`

Any thoughts?

**Comment by docwilmot** [comment]
Any thoughts anyone?

**Comment by docwilmot** [comment]
Pushed a simpler concept, just check relationships *after* contexts have loaded data.

**Comment by docwilmot** [comment]
Tsts pass, willing to advocate, can I get review or milestone please?

**Comment by argiepiano** [comment]
@docwilmot I've tested the second PR with your directions above and it works for me. I don't know enough about Layouts to do a thorough code review, but it looks good to me.

**Comment by argiepiano** [comment]
While this patch worked in earlier versions of Backdrop, after #5483 was merged into Backdrop 1.22.0, this patch stopped working as expected. Marking as needs work, but I suspect we need to revisit #5483.

**Comment by argiepiano** [comment]
New PR submitted https://github.com/backdrop/backdrop/pull/4080.

Ready for testing and review.

**Comment by yorkshire-pudding** [comment]
Tested and confirmed that this PR fixes the issue described.  I used the steps in #5636 to reproduce as was easier to follow in the UI than the steps above.

While the code looks ok, I don't understand enough of what it is doing to do a code review

**Comment by argiepiano** [comment]
Thanks @yorkshire-pudding. The code is tricky because it's part of layouts, which are pretty complex. Most of the changes were suggested by @docwilmot - I'd love it if he could review the code - @docwilmot, pretty please? 🙏🏽

**Comment by argiepiano** [comment]
Hi @**docwilmot**. I've been digging a bit into Layouts to address the issue of the contexts obtained through layout relationships not being loaded correctly.  I'd be interested in bouncing some ideas with you to see if we can find a better solution beyond the existing PR . 

Say, you create a layout with path `node/%` and a relationship to author.

Of particular interest are the function `layout_get_layout_by_path()` and the method `Layout::getContexts()`. After the first call to `Layout::getContext()` in line 1340, the function proceeds to set the context's data with the Node object provided in the $router_item.  

The issue, as you know, is that the other context object provided by the relationship (`user` in this case) is ALSO loaded within `Layout::getContexts()`, with a call to `Layout::getContextsFromRelationships()` that appears within `getContexts()`. 

During the first call to `getContexts()`, the node context doesn't yet have any data. The patch you provided, which skips calling `Layout::getContextsFromRelationships()`, and then provides a second call to `getContexts()` was intended to fix this issue. Like you, I think this approach seems a bit hacky and inelegant. 

I'm wondering if it may be better to add $router_item as an additional property of the Layout object, and then let `Layout::getContexts()` deal with setting the data for the node context before proceeding to calling `Layout::getContextsFromRelationships()`.

I'm puzzled by the fact that the node context data is initially set so late, outside  getContexts(). I think moving the data setting inside the method is more organic and may solve this issue. I noticed that there is already some data setting within the method, in line 802-810 - which perhaps could be modified to use the passed $router_item to set the data too.

Layout is complex, so maybe I'm missing important details. I realize that `Layout::getContexts()` is called everywhere, e.g. during the layout creation admin form, and during rendering, and there may be times when contexts need to be reset. So perhaps, in addition to the changes I mentioned above, there would be a need for a method that resets contexts as needed by the admin form and/or renderers.

**Comment by argiepiano** [comment]
I can give that a try and see where it leads me. Alternatively we could set the $router_item as a property of the layout object. The nice thing about that is that the layout object stays "alive" during the whole page request, and the router item will then stay alive and easily accessible from within the class at nay given moment (for example for the renderer to access if needed).

**Comment by docwilmot** [comment]
I had missed that idea. Not sure about it conceptually. The layout object would have a path and a router item? Seems like duplication somehow.

**Comment by argiepiano** [comment]
@docwilmot, I just submitted a PR with the changes I suggested. I've done some testing and it seems to be working as expected.

First, a quick response: 

> The layout object would have a path and a router item? Seems like duplication somehow.

I've checked the Layout object. It has 2 (now 3) properties that are somewhat related but not the same:
- `path`. This property is the path with the placeholder, and is only present for layout paths that have placeholders, like `node/%` of `taxonomy/term/%` or 'anything/%/some`
- `menu_item`. This property holds a `LayoutMenuItem` object, and is present when the layout path doesn't override an existing one

And the new passed property is the `router_item` array. Another option would have been to pass `href` instead (as in `node/5`). This would mean that, in order to set the node context you would have to invoke `menu_get_item(HREF)` at some point within the class, anyway. So, passing the router_item array seemed like a better solution.

The PR does the following:

- Removes the double call to `getContexts()`
- Moves the code that sets the context data using the object stored in the router item out of  `layout_get_layout_by_path()` into a new method, `EntityLayoutContext::setDataFromRouterItem()`. I believe this method is only needed by EntityLayoutContext since the `%`. Let me know if this is not the case.
- Adds a line of code to `layout_get_layout_by_path()` to pass the $router_item array as a property of the Layout objects, so that it can be accessed by methods within the class
- Adds code to set the context data by invoking `EntityLayoutContext::setDataFromRouterItem()`  in `Layout::getContexts()` so that the object is fully loaded into the context BEFORE the call to `Layout::getContextsFromRelationships()`

This is a more organic way to approach this issue, rather than the previous attempt that selectively called getContexts() twice.  Let me know what you think.

**Comment by argiepiano** [comment]
And... just noticed that some tests are not passing. I'll need to check on this tomorrow.

**Comment by argiepiano** [comment]
@docwilmot, I just pushed a new commit to PR https://github.com/backdrop/backdrop/pull/4083

A summary of what I did here:
- Same approach as my comment above
  - Moved code from `layout_get_layout_by_path()` that set the context data, into a new method context method - `setDataFromRouterItem()`
  - Unlike above, this method is now overridden differently in `EntityLayoutContext` and `LayoutStringContext`, basically doing the same thing that the block of code in `layout_get_layout_by_path()` did
- As above, I pass the `$router_item` array to the Layout object, BUT I also created a declaration of the `$router_item` property, which streamlines the code as I don't have to play with `ifs` to check if that property is set
- And finally, in order to allow the failing test to pass, also added a call to the new method `setDataFromRouterItem()` when the layout has a menu item property (i.e. when it's its own page, rather than overriding an existing path)  inside `Layout::getContexts()`

All tests are passing now, and I did a number of different manual tests with entity contexts, arguments, custo blocks that output properties of the passed contexts and strings, and string passthroughs. I would like to test with mini-layouts and other submodules to be sure this won't break any of that.

I feel good about this approach -  it's more streamlined and elegant than calling getContext twice and reseting the static cache. This approach fully loads the context data before getting to any of the relationship contexts. 

Let me know what you think.

**Comment by argiepiano** [comment]
Regardless of what approach we end up adopting, we need to create a test for contexts provided by a relationship. Apparently the existing test did not detect the wrong context data. I'll wait until my last commit is looked and commented to add a test.

**Comment by quicksketch** [comment]
Thanks @argiepiano for taking a look at this problem, and sorry to both you and @docwilmot for my slow response. I looked over the PR and it makes sense to me. I suggested some clean-up in a review: https://github.com/backdrop/backdrop/pull/4083#pullrequestreview-996053316

> And the new passed property is the router_item array. Another option would have been to pass href instead (as in node/5).

I'm slightly inclined to prefer this approach if only because debugging a `Layout` object will be simpler and we'll have less redundant data in the object. Putting the entire `router_item` in there means any entity objects will be present both under `router_item` and `context` properties.

However, we'd have to somehow explain the difference between `path` which contains wildcards and `href` which does not. (I might also call it `menu_path` rather than `href`).

**Comment by argiepiano** [comment]
Thanks, @quicksketch.

I'd like to respond to your suggestion. 

Bottom line: passing `router_item` is **much, much more efficient** and clean way to do this. This is why:

I wrote the line you quoted in your response before I learned about how the layout works for "non-overridden" paths - for example a path like `sample/%`.  It's possible for layouts with such a path to have an entity (e.g. `Node`)  as the context referred by the `%` in the path. 

So, passing `router_item` is a much preferred and direct method of getting the entity referred by `%` in those paths. For paths like `sample/%`, the `router_item` array will contain the needed entity (the `Node` in this case).

If we pass the href (e.g.  `sample/3`) we have to do a lot of work to get the `Node`. By `router_item` we get direct access to that `Node` entity.

I'll take a look at your comments later tonight and will be in touch if I have questions.

In them meantime, I would really prefer to keep the approach I developed for the PR - that is, passing `router_item` to the Layout, which includes all the needed stuff to create the context. Let me know if this is problematic to you.

**Comment by argiepiano** [comment]
Furthermore, if the custom path is something like `foo/%/bar/%` and the first `%` is a Node, and the second is a User, the `router_item` array will already have both entities in it. This is true even if one of the `%` is a passthrough string: `router_item` will have that too.

**Comment by argiepiano** [comment]
Jus thought of another possibility. Instead of passing the whole `router_item` array, which is heavy and bloated, we could just pass the value of the `map` key, which is an array of data keyed by position in the path. 

In the case of my example path above, `sample/%` where the data referred by the `%` in position 1 is a node, the `map` array would look like:
```
array(
0 => 'sample'
1 => [Node object]
```

This would make it lighter and easier to debug, which was one of your concerns.

**Comment by argiepiano** [comment]
Just a quick update: I'm done with the adjustments @quicksketch suggested, but before pushing the changes I wanted to find out why the test `testRelationships()` did not detect the problem of the wrong context data being passed.

I have figured that out. The test uses a custom block called `Layout bar block` that uses an `Author of content` relationship. That block supposedly prints the name of the user being passed to check if it corresponds to the author of the node. 

The problem reported in this issue DOES exist in the test - the context created by the relationship contains a node entity instead of a user entity (just as reported by @docwilmot  in this issue). BUT the test fails to detect this because the node entity being passed DOES contain a `name` property! So, the block prints the `name` property of the node, which corresponds to the `name` property of the user that created that node. 
 BTW this user `name` property is added to the nodes by the hook `user_node_load()`. So, the test ends up printing the correct `name` - except that it's obtaining it from the node entity, not the user entity.

As a solution, I'm proposing that instead of having the block print the user `name`, it prints a unique user property  such as the user `mail`. I'll do that tomorrow.

**Comment by argiepiano** [comment]
I just pushed changes suggested by @quicksketch.

A summary of what this PR does:

- `layout_get_layout_by_path()` passes the `$router_item['map']` to the Layout objects. Unlike the earlier version of the PR, only the much lighter `map` is passed. This map contains all the data needed for the Layout contexts

- Created the method, `LayoutContext::setDataFromRouterMap()` that sets the data of the context. It receives the router map. Unlike the earlier version of the PR, this method is not overridden by classes that extend `LayoutContext`. Rather, the class `LayoutContext` takes care of all the setting the context data

- Modified `layout.test` and `layout_test.module` to make it sensitive to the wrong object being passed in a layout relationship. The current test `testRelationships()` did not detect this (see explanation in my comment above). The changes assure that this is detected. Basically, the block `bar` now prints the user's email instead of the user's name, since `name` is available in both the `user` and the `node` objects.

I will respond to a couple of @quicksetch's comments in the PR itself.

Please review and test.

**Comment by argiepiano** [comment]
@docwilmot, @quicksketch  - wondering if you can take a look at the new PR I pushed 9 days ago? Is this better, @quicksketch? I followed your suggestions.

**Comment by argiepiano** [comment]
Any chance @docwilmot or @quicksketch (or anyone else) can do a review of this and get this serious bug fixed for 1.22.1?   Pretty please?  🙏🏽

**Comment by argiepiano** [comment]
In response to the comment from @quicksketch on the weekly dev meeting from June 23, 2022 available on youtube:

@quicksketch said:  
- "[The PR] added the menu router as a property of the layout".  My response: The latest commit adds the router item MAP, not the whole router item. This makes it a lot lighter and less complex
- "It may be possible to work with the properties we have already... I'm thinking the context information we have is already in the layout already". My response: This is at the center of the issue. The layout does NOT have the objects. The code as it is right now adds those objects (e.g. the node from the path placeholder) to the layout context within `layout_get_layout_by_path()`. The problem we are trying to solve is that these objects are added too late. So, by passing the router map (which contains the objects) to the Layout object, we are giving the Layout object to opportunity to add those objects to its Contexts objects. 

Of course I'm open to suggestions here. Initially, docwilmot (in an earlier PR that is now closed) had inserted a second call to `$layout->getContexts()`  within function `layout_get_layout_by_path ()`. The goal of the second call was to add the "author of node" user context after the node had been added. But this approach seemed hacky. The latest PR takes a different approach: providing all the information (in the router map) to let the Layout object take care of setting all the contexts (node and user) at once.

**Comment by argiepiano** [comment]
@quicksketch I believe I have figured out a way to avoid creating a new property in the Layout class - I'll work on this in the next couple of days and will report here if I'm successful.

**Comment by quicksketch** [comment]
Thanks @argiepiano! That would be amazing! And sorry for the delay on feedback here. I was hoping to provide a more concrete solution than hand-waving: "this _should_ be possible".

**Comment by argiepiano** [comment]
I've done additional research and thinking. @quicksketch, here's an update. 

Bottom line: the Layout objects don't "know" the href path of the page request (node/1 or foo/3 etc). They only "know" the internal path (node/%, foo/% etc), stored in the property `path`. This makes it difficult for my approach to work without  passing to them the router map (which contains the objects). 

## The approach of my latest PR: 
- Pass the "router map" (a small and light array that's part the router_item structure, which contains the needed objects from the path placeholders). In my PR this array is passed as a property of the Layout object
- Have the Layout object pass that map to its children context objects, by invoking the `Context::setDataFromRouterMap()` where the children set their needed data (user, node, string, whatever) 

Given the "bottom line" above, this seemed to me like the best approach.

## Alternative 1
My current PR

## Alternative 2
Remove the new $router_map property and:
1. Instead of passing the $router_map as a property to Layout, pass `$href_path` as a new property (a string) which will contain the path that originated the page request (e.g. node/1, foo/2/bar/4)
2. Within `LAyout:getContext()` obtain the full router item array by invoking `menu_get_item($href_path)`
3. Use that router item to pass the router map to  `Context::setDataFromRouterMap()` to set the data

## Alternative 3
Remove the new $router_map property and:
1. Within `Layout::getContexts()`, have the Layout object get the path with `current_path()`
2. Use `menu_get_item()` to obtain the full router item array
3. Get the router map from that array and pass it  to its children contexts by invoking `Context::setDataFromRouterMap()`

## Alternative 4
This will be a modified version of @docwilmot's closed [PR](https://github.com/backdrop/backdrop/pull/3937). That is, make two calls to `Layout::getContexts()` as follows
1. Set the data of the Layout context children OUTSIDE the Layout object. This is what is happening in 1.22.0. The context data is set outside the Layout object, in lines 1341-1373 of `layout_get_layout_by_path()`
2. Do a second call to `Layout::getContexts()` once that data has been set, BUT (unlike @docwilmot's PR), RESET the static `getContext` cache before making that call (otherwise the PR doesn't work)

Of the 4 alternatives, I still like 1 the best, and 3 the second best. I believe 4 is "hacky", and "2 is just too risky, as I may be overlooking something by using current_path()

I still think this is a serious bug, and sincerely hope it will get fixed.

**Comment by quicksketch** [comment]
Thanks @argiepiano. I spent some time trying to figure out solutions as well. 

I tried out something similar to option 4: where the static cache is reset after setting the menu contexts. I tried to file a PR against your PR but GitHub wasn't letting me, so I filed a whole new one with one extra commit at https://github.com/backdrop/backdrop/pull/4105 for you to review.

I think what I would have preferred is that we could ask for the menu contexts and _not_ set the cache for the other types of contexts. I'm not thrilled with this approach overall but it does solve the problem introduced by https://github.com/backdrop/backdrop-issues/issues/5483.

**Comment by argiepiano** [comment]
Thanks, @quicksketch. Nice. Kind of a hybrid of what I did in Approach 1 and what I suggested in 4 (which was suggested by @docwilmot). 

I've reviewed it thoroughly, and traced the whole process carefully many times, with different types of overridden and page paths and contexts, and author relationships.  

I wanted to point out that the author relationship context and its data are set as a **byproduct** of calling `$layout->checkAccess()` in line 1346.  The method `checkAccess` actually calls `Layout::getContexts()` again, and at this point the relationship author context is added to the layout object, and its data is set.

I don't know if this is what you actually intended. To me, it looks a little suspect that the author relationship context is added as a byproduct of calling something else (checkAccess)... but I guess it works, so I'm good with that.

So I will mark this RTBC. In the meantime, maybe @docwilmot may have other comments and thoughts. If not, I'd say we are good to go.

**Comment by docwilmot** [comment]
< To me, it looks a little suspect that the author relationship context is added as a byproduct of calling something else (checkAccess)

@argiepiano I agree with this. I see @quicksketch PR moves the context-object-setting into the context class but really just does what a second `getContext()` does. It would be better I think if calling `getContext()` on any layout was able to load all its contexts, including relationship contexts as well, for which ideally we'd need the router data. So I think an idea like your number 3 above would seem tidier TBH.

**Comment by argiepiano** [comment]
I just submitted a comment about the change in static cache name from `__FUNCTION__` to `__METHOD__`. I know of at least one call to `backdrop_cache_reset()` that used the function name rather than the method name - that was just merged at https://github.com/backdrop/backdrop/pull/4085. 

I suggest adding changes to those reset calls in this PR. Changing to "needs work" for that.

**Comment by argiepiano** [comment]
See #5679 and PR https://github.com/backdrop/backdrop/pull/4131. That PR should be merged to fix the static cache name change created in the PR linked to this issue (https://github.com/backdrop/backdrop/pull/4105)
